src/core/Akka/Util/ByteString.cs

```diff
@@ -28,6 +29,46 @@ namespace Akka.IO
     [DebuggerDisplay("(Count = {_count}, Buffers = {_buffers})")]
     public sealed class ByteString : IEquatable<ByteString>, IEnumerable<byte>
     {
+        public class ByteStringReadOnlySequenceSegment : ReadOnlySequenceSegment<byte>
+        {
+            public ByteStringReadOnlySequenceSegment(ReadOnlyMemory<byte> memory, long runningIndex)
+            {
+                Memory = memory;
+                RunningIndex = runningIndex;
+            }
+            
+            /// <remarks>
+            /// This is here because <see cref="ReadOnlySequenceSegment{T}"/>
+            /// has predefined properties with Protected Setters.
+            /// </remarks>
+            public static ReadOnlySequence<byte> CreateSequence(ByteString bs)
+            {
+                var bArr = bs._buffers;
+                var first = new ByteStringReadOnlySequenceSegment(bArr[0],0);
+                ByteBuffer item = default;
+                ByteStringReadOnlySequenceSegment last = first;
+                if (bArr.Length == 2)
+                {
+                    item = bArr[1];
+                    last = new ByteStringReadOnlySequenceSegment(item, bs.Count-item.Count);
+                    first.Next = last;
+                }
+                else
+                {
+                    var prior = first;
+                    for (int i = 1; i < bArr.Length; i++)
+                    {
+                        item = bArr[i];
+                        var curr = new ByteStringReadOnlySequenceSegment(item, prior.RunningIndex+prior.Memory.Length);
+                        prior.Next = curr;
+                        prior = curr;
+                    }
+                    last = prior;
+                }
+                return new ReadOnlySequence<byte>(first,0,last,last.Memory.Length);
```

Comment:
IDK if I like this less or more than just 'going backwards on the array' and just building the Segments from last to first.

Happy for feedback (I suppose main thing being the pattern here vs setting next explicitly in ctor could impact JIT optimizations... but probably overthinking things a bit.)